src/runtime/synchronization_common.h

```diff
@@ -954,120 +875,96 @@ class fast_mutex {
         : cond_state(cond_state), mutex(mutex) {
-        unpark = signal_parking_control_unpark;
     }
-};
-
-WEAK uintptr_t signal_parking_control_unpark(void *control, int unparked, bool more_waiters) {
-    signal_parking_control *signal_control = (signal_parking_control *)control;
 
-    if (!more_waiters) {
-        uintptr_t val = 0;
-        atomic_store_relaxed(signal_control->cond_state, &val);
-    }
+    uintptr_t unpark(int unparked, bool more_waiters) override {
+        if (!more_waiters) {
+            uintptr_t val = 0;
+            atomic_store_relaxed(cond_state, &val);
+        }
 
 #if 0  // TODO: figure out why this was here.
-    return (uintptr_t)signal_control->mutex;
+        return (uintptr_t)mutex;
 #else
-    return 0;
+        return 0;
 #endif
-}
-WEAK bool broadcast_parking_control_validate(void *control, validate_action &action);
-WEAK void broadcast_parking_control_requeue_callback(void *control, const validate_action &action,
-                                                     bool one_to_wake, bool some_requeued);
-struct broadcast_parking_control : parking_control {
-    uintptr_t *cond_state;
-    fast_mutex *mutex;
-
-    ALWAYS_INLINE broadcast_parking_control(uintptr_t *cond_state, fast_mutex *mutex)
-        : cond_state(cond_state), mutex(mutex) {
-        validate = broadcast_parking_control_validate;
-        requeue_callback = broadcast_parking_control_requeue_callback;
     }
 };
 
-WEAK bool broadcast_parking_control_validate(void *control, validate_action &action) {
-    broadcast_parking_control *broadcast_control = (broadcast_parking_control *)control;
+struct broadcast_parking_control : public parking_control {
+    uintptr_t *const cond_state;
+    fast_mutex *const mutex;
 
-    uintptr_t val;
-    atomic_load_relaxed(broadcast_control->cond_state, &val);
-    // By the time this broadcast locked everything and was processed, the cond
-    // has progressed to a new mutex, do nothing since any waiting threads have
-    // to be waiting on what is effectively a different condition.
-    if (val != (uintptr_t)broadcast_control->mutex) {
-        return false;
+    ALWAYS_INLINE broadcast_parking_control(uintptr_t *cond_state, fast_mutex *mutex)
+        : cond_state(cond_state), mutex(mutex) {
     }
-    // Clear the cond's connection to the mutex as all waiting threads are going to reque onto the mutex.
-    val = 0;
-    atomic_store_relaxed(broadcast_control->cond_state, &val);
 
-    action.unpark_one = !broadcast_control->mutex->make_parked_if_locked();
-
-    return true;
-}
-
-WEAK void broadcast_parking_control_requeue_callback(void *control, const validate_action &action, bool one_to_wake, bool some_requeued) {
-    broadcast_parking_control *broadcast_control = (broadcast_parking_control *)control;
-
-    if (action.unpark_one && some_requeued) {
-        broadcast_control->mutex->make_parked();
+    bool validate(validate_action &action) override {
```

Comment:
Mark these overrides final? Maybe they can be devirtualized sometimes? If they can be devirtualized completely, maybe we should use a different implementation (e.g. non-virtual template CRTP thing)

Reply:
Added the `final`. CRTP could be worth a try.

Reply:
FWIW, `-O2` enables `-fdevirtualize`; maybe this is happening already.

Reply:
I think `-fdevirtualize` needs `final` to work, unless the compiler can prove there are no further derived classes in this case, but that seems unlikely...

Reply:
Eh, not sure that CRTP would pay for itself here; we'd be able to dispense with an indirect function call (via vtable) but we'd end up with nontrivial code duplication (the bodies of park, unpark_one, and unpark_requeue). Since we were already doing indirect function calls (via function pointer) the performance here should be a wash (just cleaner code IMHO).